From 8fb0a13de183743d3c23b563243dce6d142ea55a Mon Sep 17 00:00:00 2001
From: Frank Tang <ftang@chromium.org>
Date: Tue, 04 Feb 2025 17:16:24 -0800
Subject: [PATCH] Merged: Fix out of bound string access

Bug: 386857213
(cherry picked from commit 0242cac4b20305b03b74c2e9588003378eebeb77)

Change-Id: Ieaad3d305855131e0f1f61ee18f1d9cd6310150d
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/6242398
Commit-Queue: Shu-yu Guo <syg@chromium.org>
Reviewed-by: Deepti Gandluri <gdeepti@chromium.org>
Auto-Submit: Shu-yu Guo <syg@chromium.org>
Commit-Queue: Deepti Gandluri <gdeepti@chromium.org>
Cr-Commit-Position: refs/branch-heads/13.3@{#40}
Cr-Branched-From: 41dacffe436aeb9311879cb07648f1e36609a804-refs/heads/13.3.415@{#1}
Cr-Branched-From: 3348638c0af67c885b30891a358c89a917ac9759-refs/heads/main@{#97937}
---

--- a/v8/src/objects/js-date-time-format.cc
+++ b/v8/src/objects/js-date-time-format.cc
@@ -1689,6 +1689,10 @@
   if (m0 == ':') {
     // Ignore ':'
     p++;
+    if (len == p) {
+      // Error
+      return std::nullopt;
+    }
     m0 = flat.Get(p);
   }
   if (len - p != 2) {
--- /dev/null
+++ b/v8/test/intl/regress-386857213.js
@@ -0,0 +1,5 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+assertThrows(() => new Intl.DateTimeFormat("en", {timeZone: "+09:"}), RangeError);
